File: frontend/src/types/order.ts
/**
 * Order Types and Interfaces
 */

export interface CustomerBrief {
  id: number;
  full_name: string;
  email: string;
  phone_number: string | null;
}

export interface OrderItem {
  id: number;
  order_id: number;
  product_id: number;
  product_name: string | null;
  product_sku: string | null;
  quantity: number;
  unit_price: number;
  subtotal: number;
  created_at: string;
}

export interface Order {
  id: number;
  customer_id: number;
  customer: CustomerBrief | null;
  total_amount: number;
  status: OrderStatus;
  items: OrderItem[];
  created_at: string;
  updated_at: string;
}

export interface OrderCreate {
  customer_id: number;
  items: OrderItemCreate[];
}

export interface OrderItemCreate {
  product_id: number;
  quantity: number;
}

export interface OrderUpdate {
  status?: OrderStatus;
}

export interface OrderListResponse {
  total: number;
  page: number;
  page_size: number;
  items: Order[];
}

export interface DashboardStats {
  total_products: number;
  total_customers: number;
  total_orders: number;
  low_stock_products: LowStockProduct[];
  total_revenue: number;
}

export interface LowStockProduct {
  id: number;
  name: string;
  sku: string;
  quantity: number;
}

export enum OrderStatus {
  PENDING = "pending",
  CONFIRMED = "confirmed",
  SHIPPED = "shipped",
  DELIVERED = "delivered",
  CANCELLED = "cancelled"
}

File: backend/app/schemas/order.py
"""
Order schema definitions for request/response validation.
"""

from pydantic import BaseModel, ConfigDict, Field, field_validator
from typing import Optional, List
from datetime import datetime


class CustomerBrief(BaseModel):
    """Minimal customer info embedded in order responses."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    full_name: str
    email: str
    phone_number: Optional[str] = None


class OrderItemBase(BaseModel):
    """Base order item schema."""

    product_id: int = Field(..., gt=0, description="Product ID")
    quantity: int = Field(..., gt=0, description="Quantity ordered")


class OrderItemCreate(OrderItemBase):
    """Schema for creating order items."""
    pass


class OrderItemResponse(OrderItemBase):
    """Schema for order item response."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    order_id: int
    unit_price: float
    subtotal: float
    product_name: Optional[str] = None
    product_sku: Optional[str] = None
    created_at: datetime


class OrderBase(BaseModel):
    """Base order schema."""

    customer_id: int = Field(..., gt=0, description="Customer ID")
    items: List[OrderItemCreate] = Field(..., min_length=1, description="Order items")

    @field_validator('items')
    @classmethod
    def items_not_empty(cls, v):
        if not v:
            raise ValueError('Order must contain at least one item')
        return v


class OrderCreate(OrderBase):
    """Schema for creating a new order."""
    pass


class OrderUpdate(BaseModel):
    """Schema for updating an order."""

    status: Optional[str] = Field(None, description="Order status")

    @field_validator('status')
    @classmethod
    def status_valid(cls, v):
        valid_statuses = ['pending', 'confirmed', 'shipped', 'delivered', 'cancelled']
        if v and v not in valid_statuses:
            raise ValueError(f'Status must be one of: {", ".join(valid_statuses)}')
        return v


class OrderResponse(BaseModel):
    """Schema for order response."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    customer_id: int
    customer: Optional[CustomerBrief] = None
    total_amount: float
    status: str
    items: List[OrderItemResponse]
    created_at: datetime
    updated_at: datetime


class OrderListResponse(BaseModel):
    """Schema for order list response."""

    total: int
    page: int
    page_size: int
    items: list[OrderResponse]


class DashboardStats(BaseModel):
    """Schema for dashboard statistics."""

    total_products: int
    total_customers: int
    total_orders: int
    low_stock_products: List[dict]
    total_revenue: float
